### PR TITLE
Port over changes from dev

### PR DESCRIPTION
**Pull request by jefferai**
TODO (future PRs):

* Tests for the protocol package
* Update the README

### CODE REVIEW

File: protocol/tls.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+			continue
+		}
+
+		// If we're here we found one of our known protos so break out of
+		// the for loop and finish
+		break
+	}
+
+	// Generate a server-side certificate
+	certResp, err := l.generateServerCertificatesFn(l.ctx, l.storage, serverCertsReq, l.options...)
```
Comment (jimlambrt) [inline protocol/tls.go:118]: I might be tempted to add a check for certResp != nil since it's a caller supplied func implementation.

**Review by jimlambrt — APPROVED** [review]
Really amazing work!  Feel free to merge after reviewing my comments/suggestions.

File: protocol/doc.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// Protocol provides a listener and dial function that can be used to easily
```
[the remark this answers is not in the document]
Comment (jefferai) [reply protocol/doc.go:1]: ty!

File: protocol/dialer.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+	"github.com/hashicorp/nodeenrollment"
+	nodetls "github.com/hashicorp/nodeenrollment/tls"
+	"github.com/hashicorp/nodeenrollment/types"
+	"google.golang.org/protobuf/proto"
+)
+
+// Dial returns a function suitable for dialing a connection to an
+// InterceptingListener. It takes in storage, an address, and options.
+//
+// Supported options: WithRandomReader, WithWrapper (passed thorugh to
```
Comment (jefferai) [inline protocol/dialer.go:26]:
```suggestion
// Supported options: WithRandomReader, WithWrapper (passed through to
```

Comment (jefferai) [inline protocol/dialer.go]:
```suggestion
// Supported options: WithRandomReader, WithWrapper (passed through to
```

File: common.go [hunk 4]
```diff
@@ -16,11 +16,10 @@ func IsNil(in any) bool {
 	if in == nil {
 		return true
 	}
-	switch reflect.TypeOf(in).Kind() {
-	case reflect.Ptr, reflect.Map, reflect.Array, reflect.Chan, reflect.Slice:
-		return reflect.ValueOf(in).IsNil()
+	if reflect.ValueOf(in).Kind() != reflect.Ptr {
```
[the remark this answers is not in the document]
Comment (jefferai) [reply common.go:19]: Oh I got confused. I'm not sure how this happened -- I think I updated the function in `staging` and forgot to bring it over to `dev` before creating this PR.

File: protocol/dialer.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+	tlsConn := tls.Client(nonTlsConn, tlsConfig)
+
+	return tlsConn, nil
+}
+
+// attemptFetch creates a signed fetch request and tries to perform a TLS
+// handshake, reading the resulting certificate
+func attemptFetch(ctx context.Context, nonTlsConn net.Conn, creds *types.NodeCredentials, opt ...nodeenrollment.Option) (*types.FetchNodeCredentialsResponse, error) {
+	const op = "nodeenrollment.protocol.attemptFetch"
+
```
[the remark this answers is not in the document]
Comment (jefferai) [reply protocol/dialer.go:113]: Added

File: protocol/dialer.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+	template := &x509.Certificate{
+		AuthorityKeyId: creds.CertificatePublicKeyPkix,
+		SubjectKeyId:   creds.CertificatePublicKeyPkix,
+		ExtKeyUsage: []x509.ExtKeyUsage{
+			x509.ExtKeyUsageClientAuth,
+		},
+		DNSNames:              []string{nodeenrollment.CommonDnsName},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment | x509.KeyUsageKeyAgreement | x509.KeyUsageCertSign,
+		SerialNumber:          big.NewInt(mathrand.Int63()),
+		NotBefore:             time.Now().Add(-5 * time.Minute),
```
[the remark this answers is not in the document]
Comment (jefferai) [reply protocol/dialer.go:155]: That came later and forgot to update!

File: protocol/dialer.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+		NotAfter:              time.Now().Add(5 * time.Minute),
```
[the remark this answers is not in the document]
Comment (jefferai) [reply protocol/dialer.go:156]: I'm using `-1 *` the const not before

File: protocol/listener.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+	Options []nodeenrollment.Option
+}
+
+// New creates a new listener based on the passed in listener (which should not
+// be a TLS listener) and the given base TLS configuration. This function will
+// substitute its own TLS configuration to handle the protos specific to
+// nodeenrollment. Any connection coming in that is not using those protos will
+// simply be passed through.
+//
+// The context value is not used here, but simply passed thorugh to other
```
[the remark this answers is not in the document]
Comment (jefferai) [reply protocol/listener.go:71]: I really need to get a spell checker back on, I had one before but had so many false positives I turned it off

File: protocol/tls.go (hunk 1 above)
Comment (jefferai) [reply protocol/tls.go:118]: Good idea, fixed

File: protocol/tls.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+				return nil, fmt.Errorf("(%s) error base64-decoding fetch node creds next proto value", op)
+			}
+			// Decode the proto into the request
+			req := new(types.FetchNodeCredentialsRequest)
+			if err := proto.Unmarshal(reqBytes, req); err != nil {
+				return nil, fmt.Errorf("(%s) error unmarshaling common name value: %w", op, err)
+			}
+			// This will return a response either with Authorized false and no
+			// other data or Authorized true and encrypted values
+			fetchResp, err := l.fetchCredsFn(l.ctx, l.storage, req, l.options...)
```
[the remark this answers is not in the document]
Comment (jefferai) [reply protocol/tls.go:65]: Good idea, fixed

File: protocol/tls.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+		default:
+			continue
+		}
+
+		// If we're here we found one of our known protos so break out of
+		// the for loop and finish
+		break
+	}
+
```
[the remark this answers is not in the document]
Comment (jefferai) [reply protocol/tls.go:116]: There are checks in the input for server certs req and I'm not sure that I can do any better checks here...

**Pull request merged**